diff --git a/Users/monish/src/civicrm44/CRM/Grant/BAO/Grant.php b/CRM/Grant/BAO/Grant.php
index de5356a..517953f 100644
--- a/Users/monish/src/civicrm44/CRM/Grant/BAO/Grant.php
+++ b/CRM/Grant/BAO/Grant.php
@@ -56,36 +56,95 @@ class CRM_Grant_BAO_Grant extends CRM_Grant_DAO_Grant {
    * @return array Array of event summary values
    */
   static function getGrantSummary($admin = FALSE) {
-    $query = "
-            SELECT status_id, count(id) as status_total
-            FROM civicrm_grant  GROUP BY status_id";
-
+    $query = "SELECT
+      p.id,
+      p.label, 
+      g.status_id, 
+      count(g.id) AS status_total, 
+      sum(g.amount_total) AS amount_requested,
+      sum(g.amount_granted) AS amount_granted,
+      sum(cp.amount) AS total_paid,
+      sum(g.amount_granted)/count(g.id) AS average_amount 
+      FROM civicrm_grant_program p
+      LEFT JOIN civicrm_grant g ON g.grant_program_id = p.id
+      LEFT JOIN civicrm_entity_payment ep ON ep.entity_id = g.id AND ep.entity_table = 'civicrm_grant'
+      LEFT JOIN civicrm_payment cp ON cp.id = ep.payment_id
+      WHERE g.status_id IS NOT NULL
+      GROUP BY g.grant_program_id, g.status_id WITH ROLLUP";
+    
     $dao = CRM_Core_DAO::executeQuery($query, CRM_Core_DAO::$_nullArray);
-
-    $status = array();
-    $summary = array();
-    $summary['total_grants'] = NULL;
-    $status = CRM_Core_PseudoConstant::get('CRM_Grant_DAO_Grant', 'status_id');
-
-    foreach ($status as $id => $name) {
-      $stats[$id] = array(
-        'label' => $name,
+    
+    $status = array( );
+    $summary = array( );
+    $summary['total_grants'] = $programs = NULL;
+    $summary['no_of_grants'] = NULL;
+    $querys = "SELECT
+      v.label as label,
+      v.weight as value,
+      v.value as info
+      FROM civicrm_option_value v, civicrm_option_group g
+      WHERE  v.option_group_id = g.id
+      AND  g.name = 'grant_status'
+      AND  g.is_active = 1
+      ORDER BY v.weight";
+    $daos = CRM_Core_DAO::executeQuery($querys, CRM_Core_DAO::$_nullArray);
+    while ($daos->fetch()) {
+      $status[$daos->value] = array(
+        'weight' => $daos->value,
+        'value' => $daos->info,
+        'label' => $daos->label,
         'total' => 0,
       );
     }
-
-    while ($dao->fetch()) {
-      $stats[$dao->status_id] = array(
-        'label' => $status[$dao->status_id],
-        'total' => $dao->status_total,
+    foreach ($status as $id => $name) {
+      $stats[$status[$id]['value']] = array( 
+        'label' => $name['label'],
+        'value' => $name['value'],
+        'weight' => $name['weight'],
+        'total' => 0 
       );
-      $summary['total_grants'] += $dao->status_total;
     }
-
-    $summary['per_status'] = $stats;
+    $count = 1;
+    while ($dao->fetch()) {
+      if ($dao->N == $count) {
+        $summary['total_grants']['total_requested'] = $dao->amount_requested ? CRM_Utils_Money::format($dao->amount_requested) : CRM_Utils_Money::format(0);
+        $summary['total_grants']['total_granted'] = $dao->amount_granted ? CRM_Utils_Money::format($dao->amount_granted) : CRM_Utils_Money::format(0);
+        $summary['total_grants']['total_paid'] = $dao->total_paid ? CRM_Utils_Money::format($dao->total_paid) : CRM_Utils_Money::format(0);
+        $summary['total_grants']['total_average'] = $dao->average_amount ? CRM_Utils_Money::format($dao->average_amount) : CRM_Utils_Money::format(0);
+        continue;
+      }
+      if (!empty($dao->status_id)) {
+        $programs[$dao->label][$stats[$dao->status_id]['weight']] = array( 
+          'label' => $stats[$dao->status_id]['label'],
+          'total' => $dao->status_total,
+          'value' => $stats[$dao->status_id]['value'],
+          'amount_requested' => $dao->amount_requested ? CRM_Utils_Money::format($dao->amount_requested) : CRM_Utils_Money::format(0),
+          'amount_granted' => $dao->amount_granted ? CRM_Utils_Money::format($dao->amount_granted) : CRM_Utils_Money::format(0),
+          'total_paid' => $dao->total_paid ? CRM_Utils_Money::format($dao->total_paid) : CRM_Utils_Money::format(0),
+          'average_amount' => $dao->average_amount ? CRM_Utils_Money::format($dao->average_amount) : CRM_Utils_Money::format(0),
+          'pid' => $dao->id,
+        );
+        $programs[$dao->label] = $programs[$dao->label] + array_diff_key($status, $programs[$dao->label]); //add the two arrays
+        ksort($programs[$dao->label]);
+        $summary['total_grants']['all'] = 'All';
+        $summary['no_of_grants'] += $dao->status_total;
+      } 
+      else {
+        $programs["<b>Subtotal $dao->label </b>"]['subtotal'] = array(
+          'label' => '',
+          'total' => $dao->status_total,
+          'amount_requested' => $dao->amount_requested ? CRM_Utils_Money::format($dao->amount_requested) : CRM_Utils_Money::format(0),
+          'amount_granted' => $dao->amount_granted ? CRM_Utils_Money::format($dao->amount_granted) : CRM_Utils_Money::format(0),
+          'total_paid' => $dao->total_paid ? CRM_Utils_Money::format($dao->total_paid) : CRM_Utils_Money::format(0),
+          'average_amount' => $dao->average_amount ? CRM_Utils_Money::format($dao->average_amount) : CRM_Utils_Money::format(0),
+        );
+      }
+      $count++;
+    }
+    $summary['per_status'] = $programs;
     return $summary;
   }
-
+  
   /**
    * Function to get events Summary
    *
@@ -234,7 +293,7 @@ class CRM_Grant_BAO_Grant extends CRM_Grant_DAO_Grant {
       );
     }
 
-    if (CRM_Utils_Array::value('grant', $ids)) {
+    if (CRM_Utils_Array::value('grant_id', $ids)) {
       CRM_Utils_Hook::post('edit', 'Grant', $grant->id, $grant);
     }
     else {
@@ -256,7 +315,6 @@ class CRM_Grant_BAO_Grant extends CRM_Grant_DAO_Grant {
    */
   public static function create(&$params, &$ids) {
     $transaction = new CRM_Core_Transaction();
-
     $grant = self::add($params, $ids);
 
     if (is_a($grant, 'CRM_Core_Error')) {
@@ -269,7 +327,7 @@ class CRM_Grant_BAO_Grant extends CRM_Grant_DAO_Grant {
     if (!$id) {
       $id = CRM_Utils_Array::value('contact_id', $params);
     }
-    if (!empty($params['note']) || CRM_Utils_Array::value('id', CRM_Utils_Array::value('note', $ids))) {
+    if (CRM_Utils_Array::value('note', $params) || CRM_Utils_Array::value('id', CRM_Utils_Array::value('note',$ids))) {
       $noteParams = array(
         'entity_table' => 'civicrm_grant',
         'note' => $params['note'] = $params['note'] ? $params['note'] : "null",
@@ -278,7 +336,7 @@ class CRM_Grant_BAO_Grant extends CRM_Grant_DAO_Grant {
         'modified_date' => date('Ymd'),
       );
 
-      CRM_Core_BAO_Note::add($noteParams, (array) CRM_Utils_Array::value('note', $ids));
+      CRM_Core_BAO_Note::add($noteParams, $ids['note']);
     }
     // Log the information on successful add/edit of Grant
     $logParams = array(
