diff --git a/Users/monish/src/civicrm44/CRM/Financial/Page/AJAX.php b/CRM/Financial/Page/AJAX.php
index bfb5a7e..9f4d5bb 100644
--- a/Users/monish/src/civicrm44/CRM/Financial/Page/AJAX.php
+++ b/CRM/Financial/Page/AJAX.php
@@ -87,7 +87,8 @@ class CRM_Financial_Page_AJAX {
         ) + $selectedArray;
       }
     }
-    CRM_Utils_JSON::output($elements);
+    echo json_encode($elements);
+    CRM_Utils_System::civiExit();
   }
 
   function jqFinancialRelation($config) {
@@ -106,8 +107,7 @@ class CRM_Financial_Page_AJAX {
         '3' => array(1, 9), //revenue
         '4' => array(7), //cost of sales
       );
-      $financialAccountId = CRM_Utils_Request::retrieve('_value', 'Positive', CRM_Core_DAO::$_nullObject);
-      $financialAccountTypeId = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialAccount', $financialAccountId, 'financial_account_type_id');
+      $financialAccountTypeId = CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialAccount', $_GET['_value'], 'financial_account_type_id');
       $result = CRM_Core_PseudoConstant::get('CRM_Financial_DAO_EntityFinancialAccount', 'account_relationship');
     }
 
@@ -144,7 +144,8 @@ class CRM_Financial_Page_AJAX {
         }
       }
     }
-    CRM_Utils_JSON::output($elements);
+    echo json_encode($elements);
+    CRM_Utils_System::civiExit();
   }
 
   function jqFinancialType($config) {
@@ -152,9 +153,10 @@ class CRM_Financial_Page_AJAX {
       empty($_GET['_value'])) {
       CRM_Utils_System::civiExit();
     }
-    $productId = CRM_Utils_Request::retrieve('_value', 'Positive', CRM_Core_DAO::$_nullObject);
-    $elements = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Product', $productId, 'financial_type_id');
-    CRM_Utils_JSON::output($elements);
+
+    $elements = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Product', $_GET['_value'], 'financial_type_id');
+    echo json_encode($elements);
+    CRM_Utils_System::civiExit();
   }
 
   /**
@@ -170,7 +172,7 @@ class CRM_Financial_Page_AJAX {
       }
     }
 
-    $entityID = CRM_Utils_Request::retrieve('entityID', 'Positive', CRM_Core_DAO::$_nullObject, FALSE, NULL, 'POST');
+    $entityID  = CRM_Utils_Array::value('entityID', $_POST);
     $methods = array(
       'assign' => 'addBatchEntity',
       'remove' => 'removeBatchEntity',
@@ -212,9 +214,9 @@ class CRM_Financial_Page_AJAX {
             // Update totals when closing a batch
             $params = $totals[$recordID];
           case 'reopen':
-            $status = $op == 'close' ? 'Closed' : 'Reopened';
+            $status = $op == 'close' ? 'Closed' : 'Open';
             $ids['batchID'] = $recordID;
-            $batchStatus = CRM_Core_PseudoConstant::get('CRM_Batch_DAO_Batch', 'status_id', array('labelColumn' => 'name'));
+            $batchStatus = CRM_Core_PseudoConstant::get('CRM_Batch_DAO_Batch', 'status_id');
             $params['status_id'] = CRM_Utils_Array::key($status, $batchStatus);
             $session = CRM_Core_Session::singleton();
             $params['modified_date'] = date('YmdHis');
@@ -246,7 +248,8 @@ class CRM_Financial_Page_AJAX {
         }
       }
     }
-    CRM_Utils_JSON::output($response);
+    echo json_encode($response);
+    CRM_Utils_System::civiExit();
   }
 
   static function getFinancialTransactionsList() {
@@ -276,7 +279,9 @@ class CRM_Financial_Page_AJAX {
       array(
         'civicrm_financial_trxn.payment_instrument_id as payment_method',
         'civicrm_contribution.contact_id as contact_id',
+        'civicrm_grant.contact_id as contact_id_grant',
         'civicrm_contribution.id as contributionID',
+        'civicrm_grant.id as grantID',
         'contact_a.sort_name',
         'civicrm_financial_trxn.total_amount as amount',
         'civicrm_financial_trxn.trxn_id as trxn_id',
@@ -342,6 +347,9 @@ class CRM_Financial_Page_AJAX {
       }
     }
     $financialitems = array();
+    $formLinks = CRM_Financial_Form_BatchTransaction::links();
+    $pageLinks = CRM_Financial_Page_BatchTransaction::links();
+    $formLinks['view']['url'] = $pageLinks['view']['url'] = '%%url%%';
     while ($financialItem->fetch()) {
       $row[$financialItem->id] = array();
       foreach ($columnHeader as $columnKey => $columnValue) {
@@ -370,25 +378,34 @@ class CRM_Financial_Page_AJAX {
           $row[$financialItem->id][$columnKey] = CRM_Core_OptionGroup::getLabel('contribution_status', $financialItem->$columnKey);
         }
       }
+      if ($financialItem->grantID) {
+        $entityID = $financialItem->grantID;
+        $contactID = $financialItem->contact_id_grant;
+        $url = "civicrm/contact/view/grant";
+      }
+      elseif ($financialItem->contributionID) {
+        $entityID = $financialItem->contributionID;
+        $contactID = $financialItem->contact_id;
+        $url = "civicrm/contact/view/contribution";
+      }
       if ($statusID == CRM_Core_OptionGroup::getValue('batch_status','Open')) {
         if (isset($notPresent)) {
           $js = "enableActions('x')";
           $row[$financialItem->id]['check'] = "<input type='checkbox' id='mark_x_". $financialItem->id."' name='mark_x_". $financialItem->id."' value='1' onclick={$js}></input>";
-          $row[$financialItem->id]['action'] = CRM_Core_Action::formLink(CRM_Financial_Form_BatchTransaction::links(), null, array('id' => $financialItem->id, 'contid' => $financialItem->contributionID, 'cid' => $financialItem->contact_id));
+          $row[$financialItem->id]['action'] = CRM_Core_Action::formLink($formLinks, null, array('id' => $financialItem->id, 'contid' => $entityID, 'cid' => $contactID, 'url' => $url));
         }
         else {
           $js = "enableActions('y')";
           $row[$financialItem->id]['check'] = "<input type='checkbox' id='mark_y_". $financialItem->id."' name='mark_y_". $financialItem->id."' value='1' onclick={$js}></input>";
-          $row[$financialItem->id]['action'] = CRM_Core_Action::formLink(CRM_Financial_Page_BatchTransaction::links(), null, array('id' => $financialItem->id, 'contid' => $financialItem->contributionID, 'cid' => $financialItem->contact_id));
+          $row[$financialItem->id]['action'] = CRM_Core_Action::formLink($pageLinks, null, array('id' => $financialItem->id, 'contid' => $entityID, 'cid' => $contactID, 'url' => $url));
         }
       }
       else {
         $row[$financialItem->id]['check'] = NULL;
-        $links = CRM_Financial_Page_BatchTransaction::links();
-        unset($links['remove']);
-        $row[$financialItem->id]['action'] = CRM_Core_Action::formLink($links, null, array('id' => $financialItem->id, 'contid' => $financialItem->contributionID, 'cid' => $financialItem->contact_id));
+        unset($pageLinks['remove']);
+        $row[$financialItem->id]['action'] = CRM_Core_Action::formLink($pageLinks, null, array('id' => $financialItem->id, 'contid' => $entityID, 'cid' => $contactID, 'url' => $url));
       }
-      $row[$financialItem->id]['contact_type'] = CRM_Contact_BAO_Contact_Utils::getImage(CRM_Utils_Array::value('contact_sub_type',$row[$financialItem->id]) ? CRM_Utils_Array::value('contact_sub_type',$row[$financialItem->id]) : CRM_Utils_Array::value('contact_type',$row[$financialItem->id]) ,false, $financialItem->contact_id);
+      $row[$financialItem->id]['contact_type'] = CRM_Contact_BAO_Contact_Utils::getImage(CRM_Utils_Array::value('contact_sub_type',$row[$financialItem->id]) ? CRM_Utils_Array::value('contact_sub_type',$row[$financialItem->id]) : CRM_Utils_Array::value('contact_type',$row[$financialItem->id]) ,false, $contactID);
       $financialitems = $row;
     }
 
@@ -399,7 +416,6 @@ class CRM_Financial_Page_AJAX {
         'amount', 'trxn_id', 'transaction_date', 'payment_method', 'status', 'name', 'action',
       );
 
-    header('Content-Type: application/json');
     echo CRM_Utils_JSON::encodeDataTableSelector($financialitems, $sEcho, $iTotal, $iFilteredTotal, $selectorElements);
     CRM_Utils_System::civiExit();
   }
@@ -440,7 +456,8 @@ class CRM_Financial_Page_AJAX {
     else {
       $status = array('status' => ts("This batch is configured to include only transactions using %1 payment method. If you want to include other transactions, please edit the batch first and modify the Payment Method.", array( 1 => $paymentInstrument)));
     }
-    CRM_Utils_JSON::output($status);
+    echo json_encode($status);
+    CRM_Utils_System::civiExit();
   }
 
   static function getBatchSummary() {
@@ -461,6 +478,7 @@ class CRM_Financial_Page_AJAX {
         'opened_date' => CRM_Utils_Date::customFormat($batchInfo->created_date),
       );
 
-    CRM_Utils_JSON::output($batchSummary);
+    echo json_encode($batchSummary);
+    CRM_Utils_System::civiExit();
   }
 }
