diff --git a/Users/monish/src/civicrm44/CRM/Grant/Form/Search.php b/CRM/Grant/Form/Search.php
index 95f0cd8..d135f32 100644
--- a/Users/monish/src/civicrm44/CRM/Grant/Form/Search.php
+++ b/CRM/Grant/Form/Search.php
@@ -402,7 +402,13 @@ class CRM_Grant_Form_Search extends CRM_Core_Form {
       $this->_formValues['grant_status_id'] = $status;
       $this->_defaults['grant_status_id'] = $status;
     }
-
+    $programID = CRM_Utils_Request::retrieve('pid', 'String',
+      CRM_Core_DAO::$_nullObject
+    );
+    if ($programID) {
+      $this->_formValues['grant_program_id'] = $programID;
+      $this->_defaults['grant_program_id'] = $programID;
+    }
     $cid = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
 
     if ($cid) {
